### ISSUE TITLE
AO RTSP Adapter Error - "Could not open device '/dev/v4l2-nvenc' for reading and writing. "

### ISSUE THREAD

**Issue by Nachiket-ML** [insight-platform/Savant#825]
After failing to run the template module/run.py on WSL, I was informed Savant is not supported on WSL - https://github.com/insight-platform/Savant/issues/822.

Thus, I decided to setup GPU PV on a Hyper-V Ubuntu 22.04 VM. I got the nvidia-smi command working on the VM. The command `glxinfo | grep "OpenGL renderer string"` outputs `OpenGL renderer string: D3D12 (NVIDIA GeForce RTX 4060 Laptop GPU)`. My Ubuntu kernel is 6.5.0-45-generic. The installation instructions also mentioned that I needed to dxg dkms, so I patched the linux-msft-6.6.y dxgkrnl version from WSL (https://github.com/microsoft/WSL2-Linux-Kernel/tree/linux-msft-wsl-6.6.y). 

Here are some of the instructions I followed: 
1. https://gist.github.com/OlfillasOdikno/f87a4444f00984625558dad053255ace
2. https://gist.github.com/krzys-h/e2def49966aa42bbd3316dfb794f4d6a

However, when I run the AO RTSP container from samples/template directory, I see this error: 
```
always-on-sink-1  | RuntimeError: Received error "gst-resource-error-quark: 
Could not open device '/dev/v4l2-nvenc' for reading and writing. (7)" 
from nvv4l2h264enc0. Debug info: "v4l2_calls.c(713): gst_v4l2_open (): 
/GstPipeline:pipeline0/nvv4l2h264enc:nvv4l2h264enc0:

always-on-sink-1  | system error: No such file or directory".

always-on-sink-1  |  ERROR insight::savant::deepstream::encoding          
> You have configured NVENC-accelerated encoding, but your 
device doesn't support NVENC for codec 'h264'.
```

According to the video encode/decode GPU support matrix, my GPU is supported by NVENC, so why is the RTSP container is giving me this error? 
Since I saw that NVENC is dependent on CUDA, I tried installing CUDA12.2 (both WSL 2 and ubuntu 22.04 versions), neither of which fix the error. Do I need to patch a different version of dxgkrnl from a non-WSL source?

GPU support matrix - https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new

**Comment by bwsw** [comment]
Hi. What `nvidia-smi` shows? Please list here.

**Comment by bwsw** [comment]
When you ensure that `nvidia-smi` displays proper information about GPU similar to the following listing:

```
ivan@ivan-LOQ-15IRH8:~$ nvidia-smi 
Sun Aug  4 11:31:30 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 4060 ...    Off | 00000000:01:00.0 Off |                  N/A |
| N/A   45C    P8               1W /  60W |     18MiB /  8188MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      5084      G   /usr/lib/xorg/Xorg                           14MiB |
+---------------------------------------------------------------------------------------+
```

You must configure your environment precisely as described here: https://docs.savant-ai.io/develop/

Be sure to complete all steps; do not go further if the previous step resulted in an error. Every single command must succeed. You do not need to install anything else in your host (VM) environment other than described in the document by the link.

**Comment by Nachiket-ML** [comment]
> Hi. What `nvidia-smi` shows? Please list here.

This is the output when I run `/usr/lib/wsl/lib/nvidia-smi` 

```
Sun Aug  4 14:52:03 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.27                 Driver Version: 560.70         CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4060 ...    On  |   00000000:01:00.0 Off |                  N/A |
| N/A   42C    P8              1W /   80W |      80MiB /   8188MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A       577      G   /Xorg                                       N/A      |
|    0   N/A  N/A       854      G   /unity-greeter                              N/A      |
|    0   N/A  N/A      3942      G   /Xorg                                       N/A      |
|    0   N/A  N/A      4090      G   /xdg-desktop-por                            N/A      |
+-----------------------------------------------------------------------------------------+
```

I copied the driver and lib files from my WSL VM into the Hyper-V Ubuntu 22.04 VM at `/usr/lib/wsl/drivers/` and `usr/lib/wsl/lib` locations. More info on this procedure is mentioned in the links I pasted at the top of this page. I think that there might be something wrong with this approach, since it uses the Nvidia drivers from WSL and not Ubuntu. 

Also, the docs say to install `nvidia-driver-535`, which several versions behind my up-to-date drivers (560.xx). Will that discrepancy cause any issues in Savant? Should I try installing `nvidia-driver-555` on my Hyper-V VM instead?

**Comment by bwsw** [comment]
Please follow the docs provided. As long as GPU pass-through is congigured, you just need to setup guest Ubuntu VM as described. You can use fresh drivers, in the docs, the Nvidia-recommended driver version specified.

**Comment by Nachiket-ML** [comment]
How did you manage to get the driver version showing as 535.183.01 in `nvidia-smi` output? Though I installed the Nvidia driver 535 as specified in the savant docs, I couldn't get the driver version to be 535.183.01.

**Comment by Nachiket-ML** [comment]
What kernel version are you using on the Ubuntu guest? Are you using an Ubuntu VM or do you have Ubuntu installed on the host via dual boot? I am using kernel version 6.5.0-45-generic on Hyper-V VM.

**Comment by bwsw** [comment]
> How did you manage to get the driver version showing as 535.183.01 in `nvidia-smi` output? Though I installed the Nvidia driver 535 as specified in the savant docs, I couldn't get the driver version to be 535.183.01.

You do not need exactly the same version. Nvidia changes minor version and it should not be a problem.

**Comment by bwsw** [comment]
> What kernel version are you using on the Ubuntu guest? Are you using an Ubuntu VM or do you have Ubuntu installed on the host via dual boot? I am using kernel version 6.5.0-45-generic on Hyper-V VM.

Unfortunately, we do not use Hyper-V and I cannot help with the question. I only know that our customers use Hyper-V and somehow managed to get a working VM.

Dual boot is OK.

Kernel version can vary as long as you are using Ubuntu 22.

**Comment by Nachiket-ML** [comment]
> > How did you manage to get the driver version showing as 535.183.01 in `nvidia-smi` output? Though I installed the Nvidia driver 535 as specified in the savant docs, I couldn't get the driver version to be 535.183.01.
> 
> You do not need exactly the same version. Nvidia changes minor version and it should not be a problem.

Let me rephrase - I am completely unable to change the driver version. Before installing nvidia-driver-535, `nvidia-smi` shows me driver version 560. After installing nvidia-driver-535, `nvidia-smi` still shows me version 560. The 560 version came from my WSL Nvidia drivers, not from the ubuntu driver. Do you know how to override the WSL driver with the Ubuntu driver?

**Comment by Nachiket-ML** [comment]
> > What kernel version are you using on the Ubuntu guest? Are you using an Ubuntu VM or do you have Ubuntu installed on the host via dual boot? I am using kernel version 6.5.0-45-generic on Hyper-V VM.
> 
> Unfortunately, we do not use Hyper-V and I cannot help with the question. I only know that our customers use Hyper-V and somehow managed to get a working VM.
> 
> Dual boot is OK.
> 
> Kernel version can vary as long as you are using Ubuntu 22.

Did they install the Nouveau driver by going to Nvidia website and downloading runfile? Or did they just follow the Savant installation directions and install the Ubuntu Nvidia driver? It seems to me that the runfile installation may actually update the driver version output from `nvidia-smi`.
